< [GMIT Data Analytics](https://web.archive.org/web/20201029063153/https://www.gmit.ie/computer-science-and-applied-physics/higher-diploma-science-computing-data-analytics-ict) | [Table of Contents](https://github.com/E6985) | [README](https://github.com/E6985/fda-tasks/blob/master/README.md) >

<img src="img/gmit.png" />

## Fundamentals of Data Analysis - Tasks 2020
## Due: last commit on or before December 18th, 2020
***

#### Task 1
Write a Python function called counts that takes a list as input and returns a dictionary of unique items in the list as keys and the number of times each item appears as values. So, the input ``['A', 'A', 'B', 'C', 'A']`` should have output ``{'A': 3, 'B': 1, 'C': 1}``. Your code should not depend on any module from the standard library or otherwise. By the standard library, we mean the modules and packages that come as standard with Python. Anything built-in that can be used without an import statement can be used. You should research the task first and include a description with references of your algorithm in the notebook.
***

In [1]:
nArgList=['A','A','B','C','A']

###### Dictionary

- dictionary - unorderd data collection type - flexible mappings of keys to values - form the basis of much of pythons internal implementation - created via a comma-separated list - pairs - ``key:value`` - within curly braces [[...] J. VanderPlas, "Built-In Data Structures," A Whirlwind Tour of Python, p. 60, August 2016.]

###### Module - collections

- built-in - module - collections - provides alternatives to pythons general purpose built-in container - dict [[...] Python Documentation, "Collections - Container dataypes," [Python](https://web.archive.org/web/20201022092017/https://docs.python.org/3/library/collections.html#module-collections), October 2008.]

In [2]:
# a_S_tandard_L_ibrary-imports
from collections import Counter as Cctr

- module - collections - class - Counter - elements are stored as dictionary keys - item counts are stored as dictionary values [[...] Documentation, "Collections - Container dataypes," [Python](https://web.archive.org/web/20201022092017/https://docs.python.org/3/library/collections.html#module-collections), October 2008.] - this task can depend on any module from the standard library or otherwise - ``collections`` - should not be used - the idea here is to write code from scratch - putting some decent constraints on the input - the types of things that can be in the list - reviewing module - collections - to get a feel for how it works [[...] I. McLoughlin, "Tasks 2020 Fundamentals of Data Analysis," GMIT, October 2020.]

In [3]:
# module collections - object Counter alias Cctr - instantiate
Cctr(nArgList)

Counter({'A': 3, 'B': 1, 'C': 1})

###### Multiset Theory

- implementation based on - multiset theory - shortened to - mset - multiple-membership set - common accepted name replacing - bag - bunch - weighted set - occurrence set - heap - sample - fireset - modification of - set - collection of objects called elements in which the elements can occur many times - order is ignored - number of elements occurring in a multiset is called - multiplicity [[...] Blizard, Wayne D., "Multiset theory," [Project euclid](https://web.archive.org/web/20201105163941/https://projecteuclid.org/euclid.ndjfl/1093634995), July 1987] 

- multiplicity is explicitly significant - precise number of elements [[...] Weisstein, Eric W. "Multiset," [From MathWorld--A Wolfram Web Resource](https://web.archive.org/web/20200319012240/https://mathworld.wolfram.com/Multiset.html), 2002]

$$\{3,1,2\}\equiv\{1,2,3\}$$

$$\{3,3,1,2\}\ne\{1,2,3\}$$

- the multiplicity of an element is defined as the number of times it occurs in the multiset - denoted by listing its elements separated by commas between curly braces - similiar to pythons unorderd data collection type - dictionary [[...] Jesse Nieminen, Jimin Khim, "Multiset," [Brilliant](https://web.archive.org/web/20201105164908/https://brilliant.org/wiki/multiset), April 2018.]

$$\{a,a,b,c,a\}$$

- elements listed with their multiplicities as colon-separated pairs [[...] Jesse Nieminen, Jimin Khim, "Multiset," [Brilliant](https://web.archive.org/web/20201105164908/https://brilliant.org/wiki/multiset), April 2018.]

$$\{a:3,b:1,c:1\}$$

###### Algorithm - Research - module - collections

- source code - module - collections - Class - Counter - method - most_common [[...] Python - "cpython/Lib/collections/__init__.py," [Github](https://web.archive.org/web/20201105151531/https://github.com/python/cpython/blob/master/Lib/collections/__init__.py), May 2020.] - algorithm used - programming language - Smalltalk - Bag.sortedByCount [[...] Smalltalk - " smalltalk/kernel/Bag.st," [Github](https://web.archive.org/web/20201105154907/https://github.com/gnu-smalltalk/smalltalk/blob/master/kernel/Bag.st), January 2010.] - line - 568 [[...] Python - "cpython/Lib/collections/__init__.py," [Github](https://web.archive.org/web/20201105151531/https://github.com/python/cpython/blob/master/Lib/collections/__init__.py), May 2020.]

In [4]:
%%script bash
cat -n C:/Users/E6985/anaconda3/lib/collections/__init__.py | grep "def most_common" -A 11

   559	    def most_common(self, n=None):
   560	        '''List the n most common elements and their counts from the most
   561	        common to the least.  If n is None, then list all element counts.
   562	
   563	        >>> Counter('abracadabra').most_common(3)
   564	        [('a', 5), ('b', 2), ('r', 2)]
   565	
   566	        '''
   567	        # Emulate Bag.sortedByCount from Smalltalk
   568	        if n is None:
   569	            return sorted(self.items(), key=_itemgetter(1), reverse=True)
   570	        return _heapq.nlargest(n, self.items(), key=_itemgetter(1))


###### Algorithm - Research - Object-Oriented Approach

- keyword - class - indicates creation - in python all classes create objects - implicit default - ``class (object)`` - python 3 - ``(object)`` - not required [[...] D. Amos, "Object-Oriented Programming (OOP) in Python 3," [Real Python](https://web.archive.org/web/20201107154103/https://realpython.com/python3-object-oriented-programming), July 2020.] - source code - Appendix A - object-oriented implementation of mulitset - bag - whereby - class - Counts - derives from object - dict [[...] E. Wiethoff, "Bag Sum In Many Programming Languages," [wiki.c2](https://web.archive.org/web/20201107143701/https://wiki.c2.com/?BagSumInManyProgrammingLanguages), n.d.]

```python
# define class - Counts
class Counts(dict):
```

- method - \_\_init\_\_ - initialise - specify - an objects initial attributes by giving them their default values - state - must have at least one argument - variable - self - refers to the object itself - Counts - the method - \_\_init\_\_ - called automatically when creating a new - Counts - instance [[...] D. Amos, "Object-Oriented Programming (OOP) in Python 3," [Real Python](https://web.archive.org/web/20201107154103/https://realpython.com/python3-object-oriented-programming), July 2020.]

```python
# instance attributes - initialiser
def __init__(self, nArgInput):
```

- loop through each element in the input list - method - performCount - called on each element [[...] No Reference.]

```python
    for nEachElement in nArgInput:
        self.performCount(nEachElement)
```

- method - performCount - the object - self - set of elements from the input list - assigned the count of each occurrence of the list elements - keeping count - add one to each occurrence [[...] W3Schools, "Python Dictionary get() Method," [w3schools](https://web.archive.org/web/20201107161412/https://www.w3schools.com/python/ref_dictionary_get.asp), August 2018.] 

```python
# instance attributes - initialiser
def performCount(self,nEachElement):
    self[nEachElement]=self.get(nEachElement,0)+1
```

- method __dict__ - single line - list comprehension - compressing a list-building - for - loop into a single short readable line [[...] J. VanderPlas, "Built-In Data Structures," A Whirlwind Tour of Python, p. 58, August 2016.] 


- each ``key`` - ``value`` - ìn the newly returned sorted set from the elements in the iterable - object - self - looping through the dictionary - the key plus corresponding value - retrieved at the same time using the method - items [[...] Python Documentation, "5.6. Looping Techniques," [Python](https://web.archive.org/web/20201107162523/https://docs.python.org/3/tutorial/datastructures.html#looping-techniques), October 2008.]
    
```python
def __dict__(self):
    return [(nEachKey,nEachvalue) for (nEachKey,nEachvalue) in sorted(self.items())]
```

- reviewing the implementation from the module - collections - can see similarities - line - 569 [[5] Python - "cpython/Lib/collections/__init__.py," [Github](https://web.archive.org/web/20201105151531/https://github.com/python/cpython/blob/master/Lib/collections/__init__.py), May 2020.]

In [5]:
%%script bash
cat -n C:/Users/E6985/anaconda3/lib/collections/__init__.py | grep "# Emulate Bag.sortedByCount from Smalltalk" -A 2

   567	        # Emulate Bag.sortedByCount from Smalltalk
   568	        if n is None:
   569	            return sorted(self.items(), key=_itemgetter(1), reverse=True)


In [6]:
# define class - Counts
class Counts(dict):
    # instance attributes - initialiser
    def __init__(self,nArgInput):
        for nEachElement in nArgInput:
            self.performCount(nEachElement)
    # instance attributes - initialiser
    def performCount(self,nEachElement):
        self[nEachElement]=self.get(nEachElement,0)+1
    # instance attributes - initialiser
    def __dict__(self):
        return [(nEachKey,nEachvalue) for (nEachKey,nEachvalue) in sorted(self.items())]

In [7]:
# object Counts - instantiate
Counts(nArgList)

{'A': 3, 'B': 1, 'C': 1}

###### Function counts

- input - list - ``['A', 'A', 'B', 'C', 'A']`` - function - counts - should take any list containing any items and return a dictionary of unique items in the list as keys and the number of times each item appears as values - algorithm designed should be based on imagining the list contains - letters in strings as presented in the Tasks 1 question - list containing something other than strings - containing a single character - containing numbers - containing other lists - containing dictionaries - **think about this** - ok to **specify if the the function does not work with** lists of lists or something like that but consider said ideas - consider all of the different ins and outs where the algorithm might fail and document - have a lot of freedom here [[...] I. McLoughlin, "Tasks 2020 Fundamentals of Data Analysis," GMIT, October 2020.]

In [8]:
nArgList=['A', 'A', 'B', 'C', 'A']

- function - counts - create an empty dictionary - nDictionary - will eventually be returned - class - Dict - method - setdefault - **key(s)** - list items - **value** - default set to zero for each count of each item - key - in the new dictionary - each iteration through the loop adds one to the - value via indexing - **indexing poor wording - dictionaries do not do indexing** [[...] Python Documentation, "Built-in Types," [Python](https://web.archive.org/web/20201101185543/https://docs.python.org/3/library/stdtypes.html#dict.setdefault), October 2008.]

In [9]:
# define function - counts
def counts(nParList):
    nDictionary={}
    for nEachValue in nParList:
        # object Dict - method setdefault - call
        nDictionary.setdefault(nEachValue,0)
        nDictionary[nEachValue]+=1
    return nDictionary

In [10]:
# function counts - call
counts(nArgList)

{'A': 3, 'B': 1, 'C': 1}

***

Further Research
- http://www.gnu.org/software/smalltalk/manual-base/html_node/Bag.html
- http://www.demo2s.com/Tutorial/Cpp/0380__set-multiset/Catalog0380__set-multiset.htm
- http://code.activestate.com/recipes/259174/
- https://code.activestate.com/recipes/259174/
- https://wiki.c2.com/?BagSumInManyProgrammingLanguages

***

***
#### References
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[1] authorNames, referenceTitle, locationWebsiteBookVideo, dateMonthYear.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[10]

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[100]

&nbsp;&nbsp;&nbsp;[1000]
***
#### END

#### Task 2
Write a Python function called dicerolls that simulates rolling dice. Your function should take two parameters: the number of dice ``k`` and the number of times to roll the dice ``n``. The function should simulate randomly rolling ``k`` dice ``n`` times, keeping track of each total face value. It should then return a dictionary with the number of times each possible total face value occurred. So, calling the function as ``diceroll(k=2, n=1000)`` should return a dictionary like:

``{2:19,3:50,4:82,5:112,6:135,7:174,8:133,9:114,10:75,11:70,12:36}``

You can use any module from the Python standard library you wish and you should include a description with references of your algorithm in the notebook.
***

***
#### References
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[1] authorNames, referenceTitle, locationWebsiteBookVideo, dateMonthYear.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[10]

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[100]

&nbsp;&nbsp;&nbsp;[1000]
***
#### END

#### Task 3
&#x1F4CC;
***

***
#### References
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[1] authorNames, referenceTitle, locationWebsiteBookVideo, dateMonthYear.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[10]

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[100]

&nbsp;&nbsp;&nbsp;[1000]
***
#### END

#### Task 4
&#x1F4CC;
***

***
#### References
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[1] authorNames, referenceTitle, locationWebsiteBookVideo, dateMonthYear.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[10]

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[100]

&nbsp;&nbsp;&nbsp;[1000]
***
#### END

***
#### Appendix A

```python
# "Classy" solution...
class Bag(dict):
    def __init__(self, alist):
        for elem in alist:
            self.add(elem)

    def add(self, elem):
        self[elem] = self.get(elem, 0) + 1

    def __str__(self):
        out = ['%-8s %3d' % (key, val)
                        for (key, val) in sorted(self.items())]
        return '\n'.join(out)

print Bag('foo bar bar foo glag foo'.split())
```

[[...] E. Wiethoff, "Bag Sum In Many Programming Languages," [wiki.c2](https://web.archive.org/web/20201107143701/https://wiki.c2.com/?BagSumInManyProgrammingLanguages), n.d.]
***
#### END

***
# END